File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackProcessor.java
package uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd;

import com.google.common.collect.ImmutableList;
import io.vavr.Value;
import io.vavr.control.Validation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
import uk.gov.hmcts.reform.ccd.client.model.StartEventResponse;

import java.util.List;
import java.util.Map;
import java.util.Set;
import javax.annotation.Nonnull;

import static java.lang.String.format;
import static java.util.Collections.emptyList;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.CallbackValidations.hasAScannedRecord;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.CallbackValidations.hasCaseDetails;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.CallbackValidations.hasCaseReference;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.CallbackValidations.hasCaseTypeId;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.CallbackValidations.hasJurisdiction;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.CallbackValidations.isAttachEvent;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.CallbackValidations.isAttachToCaseEvent;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.Documents.checkForDuplicatesOrElse;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.Documents.getDocumentNumbers;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.Documents.getScannedDocuments;
import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.Documents.insertNewRecords;

@Service
public class CallbackProcessor {
    private static final Logger log = LoggerFactory.getLogger(CallbackProcessor.class);
    private final CcdApi ccdApi;

    public CallbackProcessor(CcdApi ccdApi) {
        this.ccdApi = ccdApi;
    }

    @SuppressWarnings("squid:S1135")
    //TODO: RPE-822 cleanup
    public List<String> process(String eventType, String eventId, CaseDetails caseDetails) {
        return Validation
            .combine(
                isAttachEvent(eventType),
                isAttachToCaseEvent(eventId),
                hasJurisdiction(caseDetails),
                hasCaseTypeId(caseDetails),
                hasCaseReference(caseDetails),
                hasAScannedRecord(caseDetails),
                hasCaseDetails(caseDetails)
            )
            .ap(this::attachCase)
            .getOrElseGet(Value::toJavaList);
    }

    @SuppressWarnings({"squid:S1172", "squid:S1135", "squid:S1854"})
    //TODO: RPE-822 these are for the validations of the incoming request and is a WIP
    private List<String> attachCase(String theType,
                                    String anEventId,
                                    String exceptionRecordJurisdiction,
                                    String caseTypeId,
                                    String caseRef,
                                    List<Map<String, Object>> exceptionDocuments,
                                    CaseDetails exceptionRecord) {
        try {
            attachCase(exceptionRecordJurisdiction, caseRef, exceptionRecord, exceptionDocuments);
            return success();
        } catch (CallbackException e) {
            return createErrorList(e);
        }
    }

    //FOR the to do warnings
    @SuppressWarnings("squid:S1135")
    private void attachCase(String exceptionRecordJurisdiction,
                            String caseRef,
                            CaseDetails exceptionRecord,
                            List<Map<String, Object>> exceptionDocuments) {
        CaseDetails theCase = ccdApi.getCase(caseRef, exceptionRecordJurisdiction);
        List<Map<String, Object>> scannedDocuments = getScannedDocuments(theCase);

        //This is done so exception record does not change state if there is a document error
        checkForDuplicatesOrElse(
            exceptionDocuments,
            scannedDocuments,
            ids -> throwDuplicateError(caseRef, ids)
        );

        StartEventResponse event = ccdApi.startAttachScannedDocs(theCase);

        ccdApi.attachExceptionRecord(theCase,
            insertNewRecords(exceptionDocuments, scannedDocuments),
            createEventSummary(exceptionRecord, theCase),
            event);
    }

    private void throwDuplicateError(String caseRef, Set<String> duplicateIds) {
        throw new CallbackException(
            format("Document with documentIds %s is already attached to %s", duplicateIds, caseRef)
        );
    }

    private String createEventSummary(CaseDetails exceptionRecord, CaseDetails theCase) {
        return format("Attaching exception record(%d) document numbers:%s to case:%d",
            exceptionRecord.getId(),
            getDocumentNumbers(Documents.getScannedRecords(exceptionRecord.getData())),
            theCase.getId());
    }

    @Nonnull
    private List<String> createErrorList(CallbackException e) {
        String message = e.getMessage();
        log.error(message, e);
        return ImmutableList.of(message);
    }

    @Nonnull
    private List<String> success() {
        return emptyList();
    }
}

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackProcessor.kt
package uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd

import feign.FeignException
import mu.KotlinLogging
import org.springframework.stereotype.Service
import uk.gov.hmcts.reform.ccd.client.CoreCaseDataApi
import uk.gov.hmcts.reform.ccd.client.model.CallbackRequest
import uk.gov.hmcts.reform.ccd.client.model.CallbackTypes.ABOUT_TO_SUBMIT
import uk.gov.hmcts.reform.ccd.client.model.CaseDetails

private val logger = KotlinLogging.logger {}

@Service
class CallbackProcessor(
    private val authenticatorFactory: CcdAuthenticatorFactory,
    private val ccdApi: CoreCaseDataApi
) {

    /**
     * returns
     *  - a list of errors
     *  - *or* an empty list  if it successfully attaches the supplementary evidence.
     */
    fun processEventType(type: String, callback: CallbackRequest): List<String> {
        logger.info { "Processing ccd callback, type: $type, eventId: ${callback.eventId}" }

        return when (type) {
            ABOUT_TO_SUBMIT -> aboutToSubmit(callback)
            else -> internalError("invalid event: $type")
        }
    }

    private fun aboutToSubmit(callback: CallbackRequest): List<String> {
        return when (callback.eventId) {
            "attachToExistingCase" -> attachToExistingCase(callback)
            else -> internalError("Invalid event ID:${callback.eventId}")
        }
    }

    private fun attachToExistingCase(callback: CallbackRequest): List<String> {
        return if (callback.caseDetails != null) {
            handleAttachRecord(callback.caseDetails)
        } else {
            internalError("No case details supplied eventId: ${callback.eventId}")
        }
    }

    private fun handleAttachRecord(caseDetails: CaseDetails): List<String> {
        return if (caseDetails.data != null) {
            startAttachEvent(caseDetails)
        } else {
            internalError("No case details supplied eventId: ${caseDetails.id}")
        }
    }

    private fun startAttachEvent(caseDetails: CaseDetails): List<String> {
        try {
            val authenticator = authenticatorFactory.createForJurisdiction(caseDetails.jurisdiction)
            ccdApi.startEventForCaseWorker(
                authenticator.userToken,
                authenticator.serviceToken,
                authenticator.userDetails.id,
                caseDetails.jurisdiction,
                caseDetails.caseTypeId,
                caseDetails.data["attachToCaseReference"] as String?,
                "TBD"
            )
            return listOf()
        } catch (e: Exception) {
            return internalError(e)
        }
    }


    private fun internalError(error: String): List<String> {
        logger.error { error }
        return listOf("Internal Error: $error")
    }

    private fun internalError(error: Exception): List<String> {
        logger.error { error }
        return listOf("Internal Error: ${errorFromException(error)}")
    }

    private fun errorFromException(error: Exception): String {
        return when (error) {
            is FeignException -> "response ${error.status()} submitting event"
            else -> "${error::class.simpleName}:${error.message}"
        }
    }

}
